### PR TITLE
Tag Helper initializers

### PR DESCRIPTION
**Pull request by DamianEdwards**
Any `Action<ITagHelper, ViewContext>` can be registered in DI to allow setting of tag helper properties after any `[Activate]` properties are set but before their bound from the source document and the tag helper is executed
- Rolls back the previous work to add `FormTagHelperOptions` and bits to support it
- #1689 & #2162

@NTaylorMullen @dougbu @davidfowl

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Razor/ITagHelperActivator.cs [hunk 1]
```diff
@@ -13,8 +13,8 @@ public interface ITagHelperActivator
         /// <summary>
         /// When implemented in a type, activates an instantiated <see cref="ITagHelper"/>.
         /// </summary>
-        /// <param name="tagHelper">The <see cref="ITagHelper"/> to activate.</param>
+        /// <param name="tagHelper">The <see cref="TTagHelper"/> to activate.</param>
         /// <param name="context">The <see cref="ViewContext"/> for the executing view.</param>
-        void Activate(ITagHelper tagHelper, ViewContext context);
+        void Activate<TTagHelper>(TTagHelper tagHelper, ViewContext context) where TTagHelper : ITagHelper;
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/ITagHelperActivator.cs]: Need to add the typeparam to the docs.

File: src/Microsoft.AspNet.Mvc.Razor/InitializeTagHelperOfT.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNet.Razor.Runtime.TagHelpers;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Mvc.Razor
+{
+    /// <inheritdoc />
+    public class InitializeTagHelper<TTagHelper> : IInitializeTagHelper<TTagHelper>
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/InitializeTagHelperOfT.cs]: `DefaultInitializeTagHelpers`?

File: src/Microsoft.AspNet.Mvc.Razor/RazorServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -24,5 +25,25 @@ public static class ServiceCollectionExtensions
...
+        /// Adds an initialization callback for a given <see cref="ITagHelper"/> type.
+        /// </summary>
+        /// <remarks>
+        /// The callback will be invoked on the <see cref="ITagHelper"/> of the specified type before the
+        /// <see cref="ITagHelper.ProcessAsync(TagHelperContext, TagHelperOutput)"/> method is called.
+        /// </remarks>
+        /// <typeparam name="TTagHelper">The type of <see cref="ITagHelper"/> being initialized.</typeparam>
+        /// <param name="services">The services available in the application.</param>
+        /// <param name="initialize">An action to initialize the <see cref="ITagHelper"/>.</param>
+        public static void InitializeTagHelper<TTagHelper>(
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/RazorServiceCollectionExtensions.cs]: Does this need to be fluent? The `WithControllersAsServices` in MvcServiceCollectionExtensions is

File: test/Microsoft.AspNet.Mvc.Razor.Test/DefaultTagHelperActivatorTest.cs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+using Microsoft.AspNet.Http.Core;
+using Microsoft.AspNet.Mvc.ModelBinding;
+using Microsoft.AspNet.Mvc.Rendering;
+using Microsoft.AspNet.Razor.Runtime.TagHelpers;
+using Microsoft.AspNet.Routing;
+using Microsoft.Framework.DependencyInjection;
+using Moq;
+using Xunit;
+
+namespace Microsoft.AspNet.Mvc.Razor.Test
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Razor.Test/DefaultTagHelperActivatorTest.cs]: `Microsoft.AspNet.Mvc.Razor`

File: test/Microsoft.AspNet.Mvc.Razor.Test/DefaultTagHelperActivatorTest.cs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+{
+    public class DefaultTagHelperActivatorTest
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Razor.Test/DefaultTagHelperActivatorTest.cs]: Needs one test that verifies when multiple initializers are setup, all of them get called and one which verifies `InitializeTagHelper<FooTagHelper>` does not get used when activating `BarTagHelper`

File: test/Microsoft.AspNet.Mvc.Razor.Test/DefaultTagHelperActivatorTest.cs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+            Assert.NotSame(viewContext, helper.ViewContext);
+        }
+
+        private static HttpContext MakeHttpContext(IServiceProvider services = null)
+        {
+            var httpContext = new Mock<DefaultHttpContext>();
+            httpContext.CallBase = true;
+            if (services != null)
+            {
+                httpContext.Setup(c => c.RequestServices).Returns(services);
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Razor.Test/DefaultTagHelperActivatorTest.cs]: The property is settable. Why not do

```C#
var httpContext = new DefaultHttpContext();
if (services != null)
{
    httpContext.RequestServices = services;
}
```

Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Razor.Test/DefaultTagHelperActivatorTest.cs]: Plus both the things that call this method always pass in services - just drop the null check and make it non-optional.

File: src/Microsoft.AspNet.Mvc.Razor/DefaultTagHelperActivator.cs [hunk 7]
```diff
@@ -2,7 +2,10 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Collections;
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/DefaultTagHelperActivator.cs]: Is this used? Remove \ sort usings maybe?

File: src/Microsoft.AspNet.Mvc.Razor/InitializeTagHelperOfT.cs (hunk 2 above)
Comment (DamianEdwards) [inline src/Microsoft.AspNet.Mvc.Razor/InitializeTagHelperOfT.cs]: I think we've moved away from prefixing our implementations with "Default"

File: test/Microsoft.AspNet.Mvc.Razor.Test/DefaultTagHelperActivatorTest.cs (hunk 4 above)
Comment (DamianEdwards) [inline test/Microsoft.AspNet.Mvc.Razor.Test/DefaultTagHelperActivatorTest.cs]: :+1:

File: test/Microsoft.AspNet.Mvc.Razor.Test/DefaultTagHelperActivatorTest.cs (hunk 5 above)
Comment (DamianEdwards) [inline test/Microsoft.AspNet.Mvc.Razor.Test/DefaultTagHelperActivatorTest.cs]: Yep, will add

File: src/Microsoft.AspNet.Mvc.Razor/IInitializeTagHelperOfT.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+using Microsoft.AspNet.Razor.Runtime.TagHelpers;
+
+namespace Microsoft.AspNet.Mvc.Razor
+{
+    /// <summary>
+    /// Initializes an <see cref="ITagHelper"/> before it's executed.
+    /// </summary>
+    /// <typeparam name="TTagHelper">The <see cref="ITagHelper"/> type.</typeparam>
+    public interface IInitializeTagHelper<TTagHelper>
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/IInitializeTagHelperOfT.cs]: Out of curiosity - why not call this `ITagHelperInitializer`. It follows the naming convention used by other things - `IControllerActivator`, `ITagHelperActivator` etc

File: src/Microsoft.AspNet.Mvc.Razor/DefaultTagHelperActivator.cs (hunk 7 above)
Comment (DamianEdwards) [inline src/Microsoft.AspNet.Mvc.Razor/DefaultTagHelperActivator.cs]: :+1:

**Comment by pranavkm** [comment]
:shipit: once somebody else's looked at it.

File: src/Microsoft.AspNet.Mvc.Razor/IInitializeTagHelperOfT.cs (hunk 8 above)
Comment (DamianEdwards) [inline src/Microsoft.AspNet.Mvc.Razor/IInitializeTagHelperOfT.cs]: We modeled it on the `IConfigureOptions` instead.

Comment (DamianEdwards) [inline src/Microsoft.AspNet.Mvc.Razor/IInitializeTagHelperOfT.cs]: I don't feel strongly either way

File: src/Microsoft.AspNet.Mvc.Razor/RazorServiceCollectionExtensions.cs (hunk 3 above)
Comment (DamianEdwards) [inline src/Microsoft.AspNet.Mvc.Razor/RazorServiceCollectionExtensions.cs]: Yeah it should, will change.

File: src/Microsoft.AspNet.Mvc.Razor/ITagHelperActivator.cs (hunk 1 above)
Comment (DamianEdwards) [inline src/Microsoft.AspNet.Mvc.Razor/ITagHelperActivator.cs]: :+1:

File: test/Microsoft.AspNet.Mvc.Razor.Test/DefaultTagHelperActivatorTest.cs (hunk 6 above)
Comment (DamianEdwards) [inline test/Microsoft.AspNet.Mvc.Razor.Test/DefaultTagHelperActivatorTest.cs]: There is a 3rd call that doesn't pass any.

File: src/Microsoft.AspNet.Mvc.Razor/IInitializeTagHelperOfT.cs [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+        where TTagHelper : ITagHelper
+    {
+        /// <summary>
+        /// Initializes the <see cref="TTagHelper"/> using <see cref="Action"/>;
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor/IInitializeTagHelperOfT.cs]: `Initializes the <see cref="TTagHelper"/>.`

File: src/Microsoft.AspNet.Mvc.Razor/IInitializeTagHelperOfT.cs [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+        /// </summary>
+        /// <param name="helper">The <see cref="TTagHelper"/> to initialize.</param>
+        /// <param name="context">
+        ///     The <see cref="ViewContext"/> for the <see cref="IView"/> the <see cref="TTagHelper"/> is in.
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor/IInitializeTagHelperOfT.cs]: The `IView` here causes an xml-doc error doesn't it? (needs to be qualified)

File: src/Microsoft.AspNet.Mvc.Razor/IInitializeTagHelperOfT.cs [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+        /// </param>
+        void Initialize(TTagHelper helper, ViewContext context);
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor/IInitializeTagHelperOfT.cs]: I don't particularly agree with it but for consistency we've been putting `[NotNull]` attributes on interfaces and implementations.

File: src/Microsoft.AspNet.Mvc.Razor/InitializeTagHelperOfT.cs [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+        /// <param name="action">The initialization delegate.</param>
+        public InitializeTagHelper([NotNull] Action<TTagHelper, ViewContext> action)
+        {
+            Action = action;
+        }
+
+        /// <summary>
+        /// The initialization delegate.
+        /// </summary>
+        public Action<TTagHelper, ViewContext> Action { get; }
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor/InitializeTagHelperOfT.cs]: Why expose? probably better off using `private readonly Action _initializeDelegate;`

File: src/Microsoft.AspNet.Mvc.Razor/RazorServiceCollectionExtensions.cs [hunk 13]
```diff
@@ -24,5 +25,25 @@ public static class ServiceCollectionExtensions
...
+
+        /// <summary>
+        /// Adds an initialization callback for a given <see cref="ITagHelper"/> type.
+        /// </summary>
+        /// <remarks>
+        /// The callback will be invoked on the <see cref="ITagHelper"/> of the specified type before the
+        /// <see cref="ITagHelper.ProcessAsync(TagHelperContext, TagHelperOutput)"/> method is called.
+        /// </remarks>
+        /// <typeparam name="TTagHelper">The type of <see cref="ITagHelper"/> being initialized.</typeparam>
+        /// <param name="services">The services available in the application.</param>
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor/RazorServiceCollectionExtensions.cs]: `The <see cref="IServiceCollection"/> instance this method extends.`

File: src/Microsoft.AspNet.Mvc.Razor/IInitializeTagHelperOfT.cs (hunk 10 above)
Comment (DamianEdwards) [inline src/Microsoft.AspNet.Mvc.Razor/IInitializeTagHelperOfT.cs]: :+1:

File: src/Microsoft.AspNet.Mvc.Razor/IInitializeTagHelperOfT.cs (hunk 11 above)
Comment (DamianEdwards) [inline src/Microsoft.AspNet.Mvc.Razor/IInitializeTagHelperOfT.cs]: :eyes:

File: src/Microsoft.AspNet.Mvc.Razor/InitializeTagHelperOfT.cs (hunk 12 above)
Comment (DamianEdwards) [inline src/Microsoft.AspNet.Mvc.Razor/InitializeTagHelperOfT.cs]: I can't remember, probably some carry over from the initial prototype

File: test/Microsoft.AspNet.Mvc.FunctionalTests/MvcTagHelpersTest.cs [hunk 14]
```diff
@@ -10,6 +10,8 @@
 using System.Reflection;
 using System.Threading.Tasks;
 using Microsoft.AspNet.Builder;
+using Microsoft.AspNet.Mvc.TagHelpers;
+using Microsoft.AspNet.TestHost;
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/MvcTagHelpersTest.cs]: Needed?

**Comment by NTaylorMullen** [comment]
:watch:

File: test/Microsoft.AspNet.Mvc.FunctionalTests/MvcTagHelpersTest.cs (hunk 14 above)
Comment (DamianEdwards) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/MvcTagHelpersTest.cs]: Only one of them, bad merge.

**Comment by DamianEdwards** [comment]
@NTaylorMullen updated

File: src/Microsoft.AspNet.Mvc.Razor/IInitializeTagHelperOfT.cs [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+    /// <typeparam name="TTagHelper">The <see cref="ITagHelper"/> type.</typeparam>
+    public interface IInitializeTagHelper<TTagHelper>
+        where TTagHelper : ITagHelper
+    {
+        /// <summary>
+        /// Initializes the <see cref="TTagHelper"/>.
+        /// </summary>
+        /// <param name="helper">The <see cref="TTagHelper"/> to initialize.</param>
+        /// <param name="context">
+        ///     The <see cref="ViewContext"/> for the <see cref="IView"/> the <see cref="TTagHelper"/> is in.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/IInitializeTagHelperOfT.cs]: nit: we aren't consistent in how we describe `ViewContext` parameters but `The <see cref="ViewContext"/> for the executing view.` (as in `ITagHelperActivator`) seems most common.

**Comment by NTaylorMullen** [comment]
:shipit:

File: src/Microsoft.AspNet.Mvc.Razor/IInitializeTagHelperOfT.cs (hunk 8 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/IInitializeTagHelperOfT.cs]: other repos may have other conventions.  and we have a few oddballs like `IConsumesActionConstraint`, `IAllowAnonymous`, and of course `ICanHasViewContext`.  but  `ITagHelperInitializer` is much more consistent with most other interfaces we define in MVC.

File: src/Microsoft.AspNet.Mvc.Razor/ITagHelperActivator.cs [hunk 16]
```diff
@@ -13,8 +13,9 @@ public interface ITagHelperActivator
         /// <summary>
         /// When implemented in a type, activates an instantiated <see cref="ITagHelper"/>.
         /// </summary>
-        /// <param name="tagHelper">The <see cref="ITagHelper"/> to activate.</param>
+        /// <typeparam name="TTagHelper">The <see cref="ITagHelper"/> type.</typeparam>
+        /// <param name="tagHelper">The <see cref="TTagHelper"/> to activate.</param>
         /// <param name="context">The <see cref="ViewContext"/> for the executing view.</param>
-        void Activate(ITagHelper tagHelper, ViewContext context);
+        void Activate<TTagHelper>(TTagHelper tagHelper, ViewContext context) where TTagHelper : ITagHelper;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/ITagHelperActivator.cs]: nit: existing issue but should have `[NotNull]` on both parameters, matching the implementation

File: src/Microsoft.AspNet.Mvc.Razor/ITagHelperActivator.cs [hunk 17]
```diff
@@ -13,8 +13,9 @@ public interface ITagHelperActivator
         /// <summary>
         /// When implemented in a type, activates an instantiated <see cref="ITagHelper"/>.
         /// </summary>
-        /// <param name="tagHelper">The <see cref="ITagHelper"/> to activate.</param>
+        /// <typeparam name="TTagHelper">The <see cref="ITagHelper"/> type.</typeparam>
+        /// <param name="tagHelper">The <see cref="TTagHelper"/> to activate.</param>
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/ITagHelperActivator.cs]: does compiler understand this reference?  `<typeparamref name="TTagHelper"/>` would be more correct.

File: src/Microsoft.AspNet.Mvc.Razor/RazorServiceCollectionExtensions.cs [hunk 18]
```diff
@@ -24,5 +25,28 @@ public static class ServiceCollectionExtensions
         {
             services.Configure(setupAction);
         }
+
+        /// <summary>
+        /// Adds an initialization callback for a given <see cref="ITagHelper"/> type.
+        /// </summary>
+        /// <remarks>
+        /// The callback will be invoked on the <see cref="ITagHelper"/> of the specified type before the
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/RazorServiceCollectionExtensions.cs]: `the <see cref="ITagHelper"/> of the specified type` -> `any <typeparamref name="TTagHelper"> instance` or at least something that aligns better with line 30.

File: src/Microsoft.AspNet.Mvc.Razor/RazorServiceCollectionExtensions.cs [hunk 19]
```diff
@@ -24,5 +25,28 @@ public static class ServiceCollectionExtensions
         {
             services.Configure(setupAction);
         }
+
+        /// <summary>
+        /// Adds an initialization callback for a given <see cref="ITagHelper"/> type.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/RazorServiceCollectionExtensions.cs]: nit: `<see cref="ITagHelper"/>` -> `<typeparamref name="TTagHelper">` works better after "given"

File: test/Microsoft.AspNet.Mvc.Razor.Test/DefaultTagHelperActivatorTest.cs [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+            var activator = new DefaultTagHelperActivator();
+            var helper = new TestTagHelper();
+
+            // Act
+            activator.Activate(helper, viewContext);
+
+            // Assert
+            Assert.Equal(name, helper.Name);
+            Assert.Equal(number, helper.Number);
+            Assert.Same(viewDataValue, helper.ViewDataValue);
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Razor.Test/DefaultTagHelperActivatorTest.cs]: nice way to validate initializer runs after property activation

File: test/Microsoft.AspNet.Mvc.Razor.Test/DefaultTagHelperActivatorTest.cs [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+        }
+
+        [Fact]
+        public void Activate_InitializesTagHelpersAfterActivatingProperties()
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Razor.Test/DefaultTagHelperActivatorTest.cs]: nit: previous test demonstrations initializers run after activation.  this one demonstrates activation doesn't run again after initializers :smiling_imp:

File: test/Microsoft.AspNet.Mvc.Razor.Test/DefaultTagHelperActivatorTest.cs (hunk 20 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Razor.Test/DefaultTagHelperActivatorTest.cs]: nevermind, `ViewDataValue` isn't activated as I thought

File: test/Microsoft.AspNet.Mvc.Razor.Test/DefaultTagHelperActivatorTest.cs (hunk 21 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Razor.Test/DefaultTagHelperActivatorTest.cs]: also nevermind this one

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.Razor/ITagHelperActivator.cs (hunk 17 above)
Comment (DamianEdwards) [inline src/Microsoft.AspNet.Mvc.Razor/ITagHelperActivator.cs]: :+1:

**Comment by DamianEdwards** [comment]
@dougbu updated

**Comment by dougbu** [comment]
:shipit: as long as the `xml-docs-test` is clean (I may have missed something there) and after we're clear to push to dev again.

**Comment by DamianEdwards** [comment]
Pushed in 7d1c1ed8eb37f8dcb345cd2b41317cdc8789157d

**Pull request closed without merging**